OpenRA.Mods.Common/Traits/BotModules/McvExpansionManagerBotModule.cs

```diff
@@ -0,0 +1,767 @@
+				 * Attaction has a base value of indiceStrideSquare >> 3 (1/8 of the maximum distance weight, 1/(2*sqrt(2))≈ 1/2.8 of the maximum distance in map)
+				 */
+				case BotMcvExpansionMode.CheckBase:
+					var cb_conyardlocs = world.ActorsHavingTrait<Building>().Where(a => a.Owner.IsAlliedWith(player)
+						&& Info.ConstructionYardTypes.Contains(a.Info.Name)).Select(a => (a.Location, a.Owner == player)).ToArray();
+
+					CPos? cb_suitablespot = null;
+					var cb_best = int.MinValue;
+
+					foreach (var (center, value) in resourceMapIndices)
+					{
+						var attraction = indiceStrideSquare >> 3;
+
+						attraction -= (center - mcv_loc).LengthSquared / pathDistanceSquareFactor;
+
+						if (lastFailedExpandSpot == center)
+							continue;
+
+						if (world.FindActorsInCircle(world.Map.CenterOfCell(center), WDist.FromCells(resourceIndiceStride)).Any(a => !a.Disposed
+							&& (player.RelationshipWith(a.Owner) == PlayerRelationship.Enemy)
+							&& a.Info.HasTraitInfo<BuildingInfo>()
+							&& a.Info.HasTraitInfo<SellableInfo>()))
+							attraction -= indiceStrideSquare << 2;
+
+						foreach (var (location, isAlly) in cb_conyardlocs)
+						{
+							var sdistance = (center - location).LengthSquared;
+							if (sdistance <= indiceStrideSquare)
+							{
+								if (isAlly)
+									attraction -= indiceStrideSquare;
+								else
+									attraction -= indiceStrideSquare << 1;
+							}
+						}
+
+						if (!allowfallback)
+						{
+							var sdistance = (center - mcv_loc).LengthSquared;
+							if (sdistance <= indiceStrideSquare)
+								attraction -= indiceStrideSquare << 1;
+						}
+
+						if (attraction > cb_best)
+						{
+							cb_best = attraction;
+							cb_suitablespot = center;
+						}
+					}
+
+					return (cb_suitablespot ?? mcv_loc, cb_best);
+
+				/*
+				 * CheckResourceCreator mode considers the distance to current MCV, ally construction yard & refinery within range,
+				 * Attaction has a base value of indiceStrideSquare >> 4 (1/16 of the maximum distance weight, 1/4 maximum map distance in map).
+				 * Because resource amount is also a factor to be considered compared to other modes, so the base attaction is reduced by half.
+				 */
+				case BotMcvExpansionMode.CheckResource:
+
+					var cr_refinarylocs = world.ActorsHavingTrait<Refinery>().Where(a => a.Owner == player && Info.RefineryTypes.Contains(a.Info.Name))
+						.Select(a => (a.Location, a.Owner != player))
+						.ToArray();
+
+					var cr_conyardlocs = world.ActorsHavingTrait<Building>().Where(a => a.Owner.IsAlliedWith(player)
+						&& Info.ConstructionYardTypes.Contains(a.Info.Name)).Select(a => (a.Location, a.Owner != player)).ToArray();
+
+					CPos? cr_suitablespot = null;
+					var cr_best = int.MinValue;
+
+					foreach (var (center, value) in resourceMapIndices)
+					{
+						// don't have to look into cell with no resource
```

Comment:
```suggestion
						// No resources present.
```

Maybe just using better variable names would make it self-describing

Reply:
fixed